### PR TITLE
[Core][Pubsub][Logging 1/n] add logging support to GCS pubsub in Python

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
This PR adds support for publishing and subscribing to logs in Python via GCS pubsub. It also refactors the Python threaded subscriber to support subscribing and calling `close()` from multiple threads.

We can also move tests and logging support to another PR, but it will make the purpose of the refactoring seems less obvious.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/gcs_pubsub.py
Comment (ericl) [inline python/ray/_private/gcs_pubsub.py:53]: Could we move logging related code into a file like gcs_pubsub_logs.py?

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/_private/gcs_pubsub.py
Comment (mwtian) [reply python/ray/_private/gcs_pubsub.py]: Moved dict <-> LogBatch proto conversions to logging_utils.py. LMK if this is what you have in mind.

**Comment by iycheng** [comment]
Just want to confirm it, are we moving from this https://github.com/ray-project/ray/pull/20492

**Comment by mwtian** [comment]
> Just want to confirm it, are we moving from this #20492

Yeah the plan is stack #20492 after this PR. This PR adds logging to Ray pubsub. After it is merged I will merge master in #20492 and have you review the remaining parts (using Ray pubsub for logging).

**Comment by iycheng** [comment]
@mwtian  got it. I think my only question is the comment in https://github.com/ray-project/ray/pull/20492

>Thanks for this work! It looks like right now all threads are going to the same code path, including polling. I feel the code is a little bit complicated when read _poll.
Will it be better if we have an internal thread in subscriber and it's doing polling. It'll push the result to the task queue which is thread safe. The subscribe will just poll from the specific queue. will this be better?

Do you have some comments on this?

**Comment by iycheng** [comment]
@mwtian  one idea to make stacked pr clear is that we can add [x/y] at the title like this one https://github.com/ray-project/ray/pull/19679 so that we are clear about the reviewing ordering.

**Comment by mwtian** [comment]
> @mwtian got it. I think my only question is the comment in #20492
> 
> > Thanks for this work! It looks like right now all threads are going to the same code path, including polling. I feel the code is a little bit complicated when read _poll.
> > Will it be better if we have an internal thread in subscriber and it's doing polling. It'll push the result to the task queue which is thread safe. The subscribe will just poll from the specific queue. will this be better?
> 
> Do you have some comments on this?

The current implementation uses a separate thread for polling. See `_polling_loop()` / `self._polling_thread`.

File: python/ray/_private/gcs_pubsub.py
Comment (ericl) [inline python/ray/_private/gcs_pubsub.py:288]: The use of a condition variable here makes me a bit nervous--- any reason that setting the closed flag and waiting for the thread to stop is insufficient?

File: python/ray/_private/gcs_pubsub.py
Comment (ericl) [inline python/ray/_private/gcs_pubsub.py:283]: This seems pretty aggressive. Why not give up after a few retries (or not do retries)?

File: python/ray/_private/gcs_pubsub.py
Comment (ericl) [inline python/ray/_private/gcs_pubsub.py:272]: Isn't running a normal state?

**Review by ericl — COMMENTED** [review]
I'd like to see the polling loop simplified. As is, it's pretty hard to read and be confident that it's correct and behaves well under errors.

File: python/ray/_private/gcs_pubsub.py
Comment (mwtian) [reply python/ray/_private/gcs_pubsub.py]: We had to wait for either message available or subscriber closing. Refactored to not use condition anymore.

File: python/ray/_private/gcs_pubsub.py
Comment (mwtian) [reply python/ray/_private/gcs_pubsub.py]: The polling were done in a separate thread, and we wanted to have the polling thread running iff subscriber is not closed. Refactored to run polling by the poller.

**Comment by mwtian** [comment]
Refactored to use separate `GcsErrorSubscriber` and `GcsLogSubscriber` for synchronous subscription.

**Comment by iycheng** [comment]
I'm good about either. Since @ericl  has the most comment, I'll leave it to you.

File: python/ray/_private/gcs_pubsub.py
Comment (ericl) [inline python/ray/_private/gcs_pubsub.py:189]: Do we still need to worry about async closes if everything is sync? Seems like we can remove this logic and also just hold the lock during the entire poll, removing the need for the loop at all.

File: python/ray/_private/gcs_pubsub.py
Comment (ericl) [inline python/ray/_private/gcs_pubsub.py:226]: Why do we need this call?

File: python/ray/_private/gcs_pubsub.py
Comment (mwtian) [reply python/ray/_private/gcs_pubsub.py:189]: Unfortunately we still need the inner loop to check for async closes. Currently in workers, polling errors and logs are done in threads different from where subscriber.close() is called. But I simplified locking because we no longer need to lock in close().

File: python/ray/_private/gcs_pubsub.py
Comment (mwtian) [reply python/ray/_private/gcs_pubsub.py:226]: This removes the subscription state from publisher in GCS. Currently we still rely on publisher GC to remove other subscriber state, so the benefit may not be large yet. But I think it is a better practice to signal termination of the subscriber here. In future publisher can remove all related data based on this call, instead of waiting for 30~60s to GC all subscriber state. I will use a shorter deadline here to avoid slowing down close() too much.

**Review by ericl — APPROVED** [review]

**Pull request merged**
